runtime/memprof.c

```diff
@@ -156,1176 +1316,998 @@ static memprof_domain_t domain_create(caml_domain_state *caml_state)
+  update_suspended(domain, false);
 
-  if (Is_exception_result(res))
+  if (cancelled)
     caml_raise(Extract_exception(res));
 
-  /* /!\ Since the heap is in an invalid state before initialization,
-     very little heap operations are allowed until then. */
+  CAMLreturn0;
+}
+
+/**** Interface with systhread. ****/
 
-  return;
+CAMLexport memprof_thread_t caml_memprof_new_thread(caml_domain_state *state)
+{
+  CAMLassert(state->memprof);
+  return thread_create(state->memprof);
 }
 
-void caml_memprof_track_interned(header_t* block, header_t* blockend)
+CAMLexport memprof_thread_t caml_memprof_main_thread(caml_domain_state *state)
 {
-  header_t *p;
-  value callstack = 0;
-  int is_young = Is_young(Val_hp(block));
-
-  if (lambda == 0 || local->suspended) return;
-
-  p = block;
-  while (1) {
-    uintnat next_sample = rand_geom();
-    header_t *next_sample_p, *next_p;
-    if (next_sample > blockend - p)
-      break;
-    /* [next_sample_p] is the block *following* the next sampled
-       block! */
-    next_sample_p = p + next_sample;
-
-    while (1) {
-      next_p = p + Whsize_hp(p);
-      if (next_p >= next_sample_p) break;
-      p = next_p;
-    }
+  memprof_domain_t domain = state->memprof;
+  CAMLassert(domain);
+  memprof_thread_t thread = domain->threads;
+  CAMLassert(thread);
 
-    if (callstack == 0) callstack = capture_callstack_postponed();
-    if (callstack == 0) break;  /* OOM */
-    new_tracked(rand_binom(next_p - next_sample_p) + 1,
-                Wosize_hp(p), SRC_MARSHAL, is_young, Val_hp(p), callstack);
-    p = next_p;
-  }
-  check_action_pending();
+  /* There should currently be just one thread in this domain */
+  CAMLassert(thread->next == NULL);
+  return thread;
 }
 
-/**** Interface with the OCaml code. ****/
+CAMLexport void caml_memprof_delete_thread(memprof_thread_t thread)
+{
+  /* Transfer entries to the domain. If this fails due to allocation
+   * failure, we will lose the entries. */
+  (void)entries_transfer(&thread->entries, &thread->domain->entries);
+  thread_destroy(thread);
+}
 
-static void caml_memprof_init(void)
+CAMLexport void caml_memprof_enter_thread(memprof_thread_t thread)
 {
-  init = 1;
-  xoshiro_init();
+  CAMLassert(thread);
+  thread->domain->current = thread;
+  update_suspended(thread->domain, thread->suspended);
 }
 
-CAMLprim value caml_memprof_start(value lv, value szv, value tracker_param)
+/**** Interface to OCaml ****/
+
+CAMLprim value caml_memprof_start(value lv, value szv, value tracker)
 {
-  CAMLparam3(lv, szv, tracker_param);
+  CAMLparam3(lv, szv, tracker);
+  CAMLlocal1(one_log1m_lambda_v);
 
-  double l = Double_val(lv);
+  double lambda = Double_val(lv);
   intnat sz = Long_val(szv);
 
-  if (started) caml_failwith("Gc.Memprof.start: already started.");
-
-  if (sz < 0 || !(l >= 0.) || l > 1.) /* Checks that [l] is not NAN. */
+  /* Checks that [lambda] is within range (and not NaN). */
+  if (sz < 0 || lambda < 0.0 || lambda > 1.0)
```

Comment:
```suggestion
  /* Checks that [lambda] is within range (and not NaN). */
  if (sz < 0 || !(lambda >= 0.0 && lambda <= 1.0))
```

Otherwise it does not checks for NaN-ness.